File: desktop/lib/core/clawd/clawd_client.dart
import 'dart:async';
import 'dart:convert';
import 'package:web_socket_channel/web_socket_channel.dart';

// Default clawd daemon port
const int kClawdPort = 4300;

/// JSON-RPC 2.0 client for communicating with clawd over WebSocket.
///
/// Connects to ws://127.0.0.1:4300 (local) or wss://relay.clawde.io (remote).
class ClawdClient {
  ClawdClient({this.url = 'ws://127.0.0.1:$kClawdPort'});

  final String url;

  WebSocketChannel? _channel;
  int _idCounter = 0;
  final Map<int, Completer<dynamic>> _pending = {};
  final StreamController<Map<String, dynamic>> _notifications =
      StreamController.broadcast();

  bool get isConnected => _channel != null;

  /// Stream of server-push notifications (events, git status updates, etc.)
  Stream<Map<String, dynamic>> get notifications => _notifications.stream;

  Future<void> connect() async {
    _channel = WebSocketChannel.connect(Uri.parse(url));
    _channel!.stream.listen(
      _onMessage,
      onDone: _onDisconnect,
      onError: _onError,
    );
  }

  void disconnect() {
    _channel?.sink.close();
    _channel = null;
  }

  /// Send a JSON-RPC 2.0 request and return the result.
  Future<T> call<T>(String method, [Map<String, dynamic>? params]) async {
    final id = ++_idCounter;
    final completer = Completer<dynamic>();
    _pending[id] = completer;

    _channel!.sink.add(jsonEncode({
      'jsonrpc': '2.0',
      'id': id,
      'method': method,
      if (params != null) 'params': params,
    }));

    return (await completer.future) as T;
  }

  void _onMessage(dynamic raw) {
    final msg = jsonDecode(raw as String) as Map<String, dynamic>;

    if (msg.containsKey('id') && _pending.containsKey(msg['id'])) {
      // Response to a request
      final completer = _pending.remove(msg['id'])!;
      if (msg.containsKey('error')) {
        completer.completeError(ClawdError.fromJson(msg['error']));
      } else {
        completer.complete(msg['result']);
      }
    } else if (msg.containsKey('method')) {
      // Server-push notification
      _notifications.add(msg);
    }
  }

  void _onDisconnect() {
    _channel = null;
    for (final c in _pending.values) {
      c.completeError(const ClawdDisconnectedError());
    }
    _pending.clear();
  }

  void _onError(Object err) {
    _onDisconnect();
  }
}

class ClawdError implements Exception {
  const ClawdError({required this.code, required this.message});

  factory ClawdError.fromJson(Map<String, dynamic> json) {
    return ClawdError(
      code: json['code'] as int,
      message: json['message'] as String,
    );
  }

  final int code;
  final String message;

  @override
  String toString() => 'ClawdError($code): $message';
}

class ClawdDisconnectedError implements Exception {
  const ClawdDisconnectedError();

  @override
  String toString() => 'ClawdDisconnectedError: connection lost';
}

File: integrations/jetbrains/src/main/kotlin/io/clawde/ClawdClient.kt
package io.clawde

import com.intellij.openapi.Disposable
import com.intellij.openapi.diagnostic.logger
import kotlinx.serialization.json.*
import okhttp3.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

private val LOG = logger<ClawdClient>()

/**
 * Typed WebSocket/JSON-RPC 2.0 client for the clawd daemon.
 *
 * Connects to ws://127.0.0.1:4300 (default). Authenticates with a local
 * auth token read from `~/.claw/auth.token`. Each call is matched by
 * JSON-RPC `id` — callbacks are resolved on the OkHttp dispatcher thread.
 *
 * Sprint KK JB.2
 */
class ClawdClient(
    private val url: String = "ws://127.0.0.1:4300",
    private val authToken: String = readLocalAuthToken(),
) : Disposable {

    private val http = OkHttpClient()
    private var ws: WebSocket? = null
    private val idSeq = AtomicInteger(1)
    private val pending = ConcurrentHashMap<Int, (Result<JsonElement>) -> Unit>()
    private val pushListeners = mutableListOf<(method: String, params: JsonObject) -> Unit>()

    /** Connect and authenticate. Calls [onReady] when ready. */
    fun connect(onReady: () -> Unit = {}) {
        val request = Request.Builder().url(url).build()
        ws = http.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                call("daemon.auth", buildJsonObject { put("token", authToken) }) { result ->
                    result.fold(
                        onSuccess = { onReady() },
                        onFailure = { LOG.error("daemon.auth failed", it) },
                    )
                }
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handleMessage(text)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                LOG.warn("clawd WebSocket error: ${t.message}")
            }
        })
    }

    /** Send a JSON-RPC call and invoke [callback] with the result. */
    fun call(
        method: String,
        params: JsonObject = JsonObject(emptyMap()),
        callback: (Result<JsonElement>) -> Unit,
    ) {
        val id = idSeq.getAndIncrement()
        pending[id] = callback
        val msg = buildJsonObject {
            put("jsonrpc", "2.0")
            put("id", id)
            put("method", method)
            put("params", params)
        }
        ws?.send(msg.toString()) ?: callback(Result.failure(IllegalStateException("not connected")))
    }

    /** Register a listener for push notifications (method, params). */
    fun onPush(listener: (method: String, params: JsonObject) -> Unit) {
        pushListeners.add(listener)
    }

    private fun handleMessage(text: String) {
        val obj = Json.parseToJsonElement(text).jsonObject
        val id = obj["id"]?.jsonPrimitive?.intOrNull
        if (id != null) {
            val cb = pending.remove(id) ?: return
            val err = obj["error"]
            if (err != null) {
                cb(Result.failure(RuntimeException(err.jsonObject["message"]?.jsonPrimitive?.content ?: "RPC error")))
            } else {
                cb(Result.success(obj["result"] ?: JsonNull))
            }
        } else {
            // Push notification
            val method = obj["method"]?.jsonPrimitive?.content ?: return
            val params = obj["params"]?.jsonObject ?: JsonObject(emptyMap())
            pushListeners.forEach { it(method, params) }
        }
    }

    override fun dispose() {
        ws?.close(1000, "plugin disposed")
        http.dispatcher.executorService.shutdown()
    }
}

private fun readLocalAuthToken(): String {
    val home = System.getProperty("user.home") ?: return ""
    return try {
        java.io.File("$home/.claw/auth.token").readText().trim()
    } catch (_: Exception) {
        ""
    }
}
